File: RPC/2020:06:13 Junio/editorial/errorcorrection/submissions/accepted/errorcorrection_hamerly_ek_bfs.py
#!/usr/bin/env python2

import collections
import sys

# solution to errorcorrection, using bipartite matching
#
# - Create a graph like the following, with edges between words that are
#   swappable. These edges represent pairs of words that cannot be in a
#   swap-free set together.
#
# - Add a source and a sink, connecting the source to those nodes with "even"
#   permutation parity, and the sink to those with "odd" permutation parity.
#   - With this parity concept, the graph is bipartite.
#
# - Find a maximum bipartite matching using the Edmonds-Karp algorithm (Ford
#   Fulkerson with BFS).
#   - This matching represents the largest group of word pairs that cannot
#     appear together. If there were more word pairs that could be matched, the
#     matching would be larger.
#
# - The largest swap-free set is the number of original words minus the flow in
#   the network (which is the number of matched word pairs).
#
# Runtime: O(V^2 + VE^2), where V = n = number of words, and E <= V^2, so
# overall O(V^5), which is too slow. However, in reality E <= 325 * V (since
# any one word can have at most 325 = (26 * 25 / 2) edges), so it's closer to
# O(V^4) with a large constant.

def parity(s):
    '''Define the parity of a sequence as the number of inversions in the
    sequence, which we calculate brute-force. Two words with equal parity
    cannot be swappable; two words with opposing parity may be swappable. Thus
    the two parities form the two sides of our bipartite graph.'''
    p = 0
    for i, c in enumerate(s):
        for j, c2 in enumerate(s[i+1:], i+1):
            if c2 < c:
                p += 1
    return p % 2

SOURCE = 0
SINK = 1

def bfs(flow):
    '''Use BFS to find a path with remaining flow in the network from SOURCE to
    SINK, if one exists.'''

    global SOURCE, SINK

    frontier = collections.deque([SOURCE])
    back_links = {SOURCE: -1}

    while frontier and SINK not in back_links:
        current = frontier.popleft()

        for neighbor, f in flow[current].items():
            if f == 1 or neighbor in back_links:
                continue
            frontier.append(neighbor)
            back_links[neighbor] = current

    if SINK in back_links:
        path = []
        node = SINK
        while node in back_links:
            path.append(node)
            node = back_links[node]

        return list(reversed(path))

    return []
def swap(s, i, j):
    l = list(s)
    l[i], l[j] = l[j], l[i]
    return ''.join(l)

def main():
    n = int(sys.stdin.readline())
    words = [sys.stdin.readline().strip() for _ in range(n)]
    word_index = {w: i for i, w in enumerate(words)}
    # NB: source = 0, sink = 1
    edges = {i: [] for i in range(n+2)}

    # compute the parity of the permutation of each word -- e.g. the parity of
    # the number of inversions are in the string. Two words can 
    parity_words = {0: [], 1: []}
    k = len(words[0])
    for i, w in enumerate(words):
        p = parity(w)
        parity_words[p].append(i)

        # if the parity is 0, then find out if there are any one-swap neighbors
        # that are also in the word list
        if p == 0:
            for a in range(k):
                for b in range(a+1,k):
                    s = swap(w, a, b)
                    if s in word_index:
                        edges[i+2].append(word_index[s]+2) # NB: +2 for SOURCE and SINK

    for i in parity_words[0]:
        edges[SOURCE].append(i+2) # NB: +2 for SOURCE and SINK
    for i in parity_words[1]:
        edges[i+2].append(SINK)

    if 1 < len(sys.argv):
        print('digraph G {')
        for u in edges:
            if 2 <= u:
                color = 'green' if (u - 2) in parity_words[0] else 'red'
                print('{} [label={},style=filled,fillcolor={}];'.format(u, words[u-2], color))
            for v in edges[u]:
                print("{} -> {};".format(u, v))
        print('};')
        return

    flow = {u: {} for u in edges}
    for u in edges:
        for v in edges[u]:
            flow[u][v] = 0
            flow[v][u] = 1

    # Edmonds-Karp
    while True:
        path = bfs(flow)
        if not path:
            break

        for i in range(len(path) - 1):
            a, b = path[i:i+2]
            flow[a][b] += 1
            flow[b][a] -= 1

    total_flow = sum(flow[0].values())
    print(n - total_flow)

if __name__ == '__main__':
    main()

File: RPC/2020:06:13 Junio/editorial/errorcorrection/submissions/accepted/errorcorrection_hamerly_ek_bfs.cpp
#include <iostream>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>
#include <unordered_map>
#include <cassert>

int parity(std::string const &s) {
    bool p = false;
    for (auto i = s.begin(); i != s.end(); ++i)
        for (auto j = i + 1; j != s.end(); ++j)
            if (*j < *i)
                p = !p;
    return (int)p;
}

std::string swap(std::string const &s, size_t i, size_t j) {
    std::string w(s);
    w[i] = s[j];
    w[j] = s[i];
    return w;
}

int const SOURCE = 0;
int const SINK = 1;

typedef std::vector<std::vector<int>> Graph;
typedef std::vector<std::map<int, int>> Flow;

void bfs(Flow const &flow, std::vector<int> *path) {
    std::queue<int> frontier{{SOURCE}};

    std::map<int, int> back_links;
    back_links[SOURCE] = -1;

    while (frontier.size() and back_links.find(SINK) == back_links.end()) {
        int current = frontier.front();
        frontier.pop();

        for (auto [neighbor, f] : flow[current]) {
            if ((f == 1) or (back_links.find(neighbor) != back_links.end()))
                continue;
            frontier.push(neighbor);
            back_links[neighbor] = current;
        }
    }

    path->clear();
    if (back_links.find(SINK) != back_links.end()) {
        auto node = SINK;
        while (back_links.find(node) != back_links.end()) {
            path->push_back(node);
            node = back_links[node];
        }

        std::reverse(path->begin(), path->end());
    }
}

int main() {
    int n;
    std::cin >> n;

    std::vector<std::string> words(n);
    std::unordered_map<std::string, size_t> word_index;
    for (int i = 0; i < n; ++i) {
        std::cin >> words[i];
        word_index[words[i]] = i;
    }

    Graph edges(n+2);

    std::vector<std::vector<int>> parity_words(2);

    size_t k = words[0].size();
    for (int i = 0; i < n; ++i) {
        int p = parity(words[i]);
        parity_words[p].push_back(i);

        if (p == 0) {
            for (size_t a = 0; a < k; ++a) {
                for (size_t b = a + 1; b < k; ++b) {
                    auto s = swap(words[i], a, b);
                    auto si = word_index.find(s);
                    if (si != word_index.end())
                        edges[i+2].push_back(si->second+2);
                }
            }
        }
    }

    for (auto i : parity_words[0]) edges[SOURCE].push_back(i+2);
    for (auto i : parity_words[1]) edges[i+2].push_back(SINK);

    Flow flow(n+2);
    for (int i = 0; i < n + 2; ++i)
        for (auto j : edges[i]) {
            flow[i][j] = 0;
            flow[j][i] = 1; // NB: kind of a hack
        }

    std::vector<int> path;
    while (true) {
        bfs(flow, &path);
        if (path.size() == 0)
            break;

        for (size_t i = 1; i < path.size(); ++i) {
            auto a = path[i-1], b = path[i];
            flow[a][b] += 1;
            flow[b][a] -= 1;
        }
    }

    int total_flow = 0;
    for (auto [_, f] : flow[0])
        total_flow += f;
    std::cout << (n - total_flow) << std::endl;

    return 0;
}
